File: Searching/Sahil/C++/exponentialSearch.cpp
/*
Works only with sorted arrays
*/

#include<iostream>
#include<array>
using namespace std;

int binary_search(int arr[], int size, int min, int max, int number_to_search) {

    int index = -1 ;
    if( max >= 1 ) {

        while( min <= max ) {

            int mid = (max + min) / 2 ;
            
            if(arr[mid] < number_to_search) {

                min = mid + 1 ;
            } else if(arr[mid] > number_to_search) {

                max = mid - 1 ;
            } else {

                index = mid ;
                break;  
            }
        }
    }
    
    return index ;
}

int exponentialSearch(int arr[], int size, int number_to_search) {

    if (arr[0] == number_to_search) {

        return 0 ;
    } 

    int i = 1 ;
    while (i < size && arr[i] <= number_to_search) {

        i *= 2 ;
    }
    return binary_search(arr, size ,i/2, min(i, size - 1), number_to_search) ;    
}

int main() {

    array<int, 5> array = { 2, 3, 4, 10, 40 } ;

    int number_to_search = 2 ;

    int search_result = exponentialSearch(array.data(), array.size(), number_to_search) ;

    if (search_result >= 0) {

        cout << "got it at: " << search_result + 1 << endl ;
    } else {

        cout << "No such value present" << endl ;
    }
    return 0 ;
}

File: Searching/Sahil/JAVA/ExponentialSearch.java
package searching;

public class ExponentialSearch {

    public static int binarySearch(int[] arr, int low, int high, int numberToSearch) {

        int position = -1 ;
        while (low <= high) {

            int mid = (high + low) / 2 ;

            if(arr[mid] < numberToSearch) {
                low = mid + 1 ;
            } else if(arr[mid] > numberToSearch) {
                high = mid - 1 ;
            } else {
                position = mid ;
                break ;
            }
        }
        return position ;
    }

    public static int exponentialSearch(int[] arr, int numberToSearch) {

        if(arr[0] == numberToSearch) return 0 ;

        int i = 1 ;
        while(i < arr.length && arr[i] <= numberToSearch) {
            i *= 2 ;
        }
        return binarySearch(arr, i/2, Math.min(arr.length, i), numberToSearch) ;
    }

    public static void main(String[] args) {

        int[] arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200 } ;
        int numberToSearch = 10 ;
        int searchResult = exponentialSearch(arr, numberToSearch) ;

        if(searchResult >= 0) {
            System.out.println("Number found at: " + (searchResult + 1));
        } else {
            System.out.println("Number not found");
        }
    }
}
